lib/tbot/workloadidentity/crl_cache.go

```diff
@@ -0,0 +1,247 @@
+				)
+				// Set empty CRL set so consumers are unblocked.
+				m.setCRLSet(ctx, &CRLSet{})
+				return nil
+			}
+			m.logger.ErrorContext(
+				ctx,
+				"Cache failed, will attempt to re-initialize after back off",
+				"error", err,
+				"backoff", trustBundleInitFailureBackoff,
+			)
+		}
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(trustBundleInitFailureBackoff):
+			continue
+		}
+	}
+}
+
+func (m *CRLCache) watch(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	stream, err := m.revocationsClient.StreamSignedCRL(
+		ctx, &workloadidentityv1pb.StreamSignedCRLRequest{},
+	)
+	if err != nil {
+		return trace.Wrap(err, "opening CRL stream")
+	}
+
+	for {
+		res, err := stream.Recv()
+		if err != nil {
+			return trace.Wrap(err, "receiving CRL stream data")
+		}
+		m.setCRLSet(ctx, &CRLSet{
+			LocalCRL: res.Crl,
+		})
+	}
+}
+
+func (m *CRLCache) setCRLSet(ctx context.Context, crlSet *CRLSet) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	old := m.crlSet
+
+	// Exit early if the CRL set is the same as the current one.
+	if old != nil {
+		if bytes.Equal(old.LocalCRL, crlSet.LocalCRL) {
+			m.logger.DebugContext(ctx, "Ignoring unchanged CRL set")
+			return
+		}
+	}
+
+	// Clone the CRL set to avoid the caller mutating the state after it has
+	// been set.
+	m.crlSet = crlSet.Clone()
+	m.crlSet.stale = make(chan struct{})
+
+	if old == nil {
+		// Indicate that the first CRL set is now available.
+		close(m.initialized)
+	} else {
+		// Indicate that a new CRL set is available.
+		close(old.stale)
+	}
+	m.logger.DebugContext(ctx, "Broadcasting new CRL set to consumers")
+}
+
+func (m *CRLCache) getCRLSet() *CRLSet {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if m.crlSet == nil {
+		return nil
+	}
+	// Clone so a receiver cannot mutate the current state without calling
+	// setCRLSet.
+	return m.crlSet.Clone()
+}
+
+// GetCRLSet returns the current CRLSet. If the cache is not yet
+// initialized, it will block until it is.
+func (m *CRLCache) GetCRLSet(
+	ctx context.Context,
+) (*CRLSet, error) {
+	select {
+	case <-m.initialized:
+		return m.getCRLSet(), nil
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+// FetchCRLSet fetches the current CRL set from the revocations service.
+// Use this only in the implementation of OneShot methods, and prefer using the
+// cache for long-running services.
+func FetchCRLSet(
+	ctx context.Context,
+	revocationsClient workloadidentityv1pb.WorkloadIdentityRevocationServiceClient,
+) (*CRLSet, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	stream, err := revocationsClient.StreamSignedCRL(
+		ctx, &workloadidentityv1pb.StreamSignedCRLRequest{},
+	)
+	if err != nil {
+		return nil, trace.Wrap(err, "streaming CRL")
+	}
+
+	res, err := stream.Recv()
+	if err != nil {
+		if trace.IsNotImplemented(err) {
+			slog.WarnContext(ctx, "Server does not support X509 CRL functionality, no CRL will be included in the output.")
```

Comment:
Would it make sense to wrap this in a `sync.Once` or similar? I think this gets called on every renewal so it might be spammy?

Reply:
IIRC - this function is only invoked by the Workload Identity X509 output in OneShot mode, so introducing a sync.Once here wouldn't make any difference at the moment.